### ISSUE TITLE
[I] Saros and the .git folder

### ISSUE THREAD

**Issue by Janrupf** [saros-project/saros#922]
**Describe the bug**
Sharing a project containing a `.git` folder causes saros to error or destroys the `.git` folder.

**To Reproduce**
1. Create a project with a git project.
2. Try to share it.

One user will always get the following error:
```
Invitation was canceled by the remote user  because of an error on his/her side: I/O failure: IntelliJFolderImpl : .git - IntelliJProjectImpl : Project (name=<project name>, containerState=ACTIVE, componentStore=<project path>)  - Module: 'java-rs' already exists
```

After trying to reverse the sharing direction, it worked, but the git folder got deleted on my side.

**Expected behavior**
The `.git` folder should either not be shared at all, or be correctly synchronized (latter would be better!).

**Log files**
_Unable to find, apparently they moved them into a black hole since IDEA version 2020.1_

**Screenshots**
<N/A>

**Environment (please complete the following information):**
 - OS: Arch Linux
 - IDE: CLion 2020.1
 - Saros Version: commit [189a144f6f679c891350d2ec94313268aafa4c51](https://github.com/saros-project/saros/commit/189a144f6f679c891350d2ec94313268aafa4c51) (built from source to enable CLion compat)

**Comment by tobous** [comment]
Sorry for the inconvenience. I how you didn't loose any important work.

In general, Saros does not play nice when version control settings are included in the session. This is talked about in our FAQ in the section ["I’m working with a version control system, can I use Saros?"](https://www.saros-project.org/documentation/faq.html#im-working-with-a-version-control-system-can-i-use-saros).

I would recommend not sharing such directories. If you don't want to use the setup suggested in the FAQ, you can use workarounds to not share them. For Saros/I, you can mark them as `excluded` (either by right-clicking the directory and selecting `Mark Directory as > Excluded" from the menu or by setting it up in the [module configuration](https://www.jetbrains.com/help/idea/content-roots.html#folder-categories)). For Saros/E, the same can be achieved by marking the resource as ["derived"](https://help.eclipse.org/2019-12/index.jsp?topic=%2Forg.eclipse.platform.doc.isv%2Fguide%2FresAdv_derived.htm).

Note, however, that this is only a hack on our part with us hooking into an existing feature of the IDE to determine resource exclusions. This will be replaced with a full exclusion framework on our side in the future which should also support git. But that will still take a while.

> The .git folder should either not be shared at all, or be correctly synchronized (latter would be better!).

Explicitly sharing the `.git` folder is a no-go in my opinion. It contains the local configuration of the git repository, including all local branches, local settings, etc. Overwriting it for another session member could lead to them loosing a lot of work. So this is a very risky case that should not occur that often, which is why it is not worth to support it in my opinion.

Furthermore, sharing the .git folder does not make much sense either way. You can always just clone the repo yourself.
Saros does not aim to replace code management/distribution tools but rather just aims to allow you to collaborate on code. The initial synchronization is more meant to synchronize minor differences in the local project contents (i.e. you modified a file before starting the session) or to accommodate smaller/hobby projects which might not be shared through a version control system.

**Comment by tobous** [comment]
However, for now, adding a static rule to ignore all `.git` folders might be a good idea to avoid loss of data. This might restrict the usability (in case somebody is determined to share the `.git` folder or, for whatever reason, has a different, git unrelated `.git` folder in their repo), but should be worth it in my opinion. I will think about it.

**Comment by m273d15** [comment]
> However, for now, adding a static rule to ignore all .git folders might be a good idea to avoid loss of data.

Sounds good. I think there are many more users that would benefit from ignoring .git, than users that want to share it.

We should introduce it in Eclipse & IntelliJ.

**Comment by tobous** [comment]
Actually, this should not have happened with the default configuration in CLion (or JetBrains IDEs in general) as `.git` directories are ignored by default (i.e. they are part of the list of "Ignored files and folders" in `Appearance & Behavior > Editor > File Types`). As a result, they won't be detected by the filesystem and therefore also won't be shared through Saros.

Still, adding a check won't hurt.

For me, these default settings did not change when using IntelliJ 2020.1.

@Janrupf Did you manually remove the `.git` entry from the list of ignored files or does your IDE have different default settings?

Also, with the behavior described above, it sounds to me like one of the participants had the `.git` folder ignored and the other didn't. As a result, the session creation failed when the user that does not ignore the `.git` folder tries to create a session as the folder already exists on the other side but is not visible to the IDE. When trying it the other way around, Saros will detect the `.git` folder on the client side and will also detect that it does not exist on the host's side (at least not from the point of view of Saros as the directory is ignored by the IDE), meaning it will be deleted as part of the session start. The dialog will also have displayed that the folder will be deleted as part of the session negotiation and will have offered the option to cancel the negotiation to avoid this outcome.

**Comment by tobous** [comment]
And regarding the log files: Apparently JetBrains has moved them with the 2020 release according to [this support entry](https://intellij-support.jetbrains.com/hc/en-us/articles/206544519). The new system directories are listed [here](https://www.jetbrains.com/help/idea/2020.1/tuning-the-ide.html#system-directory). For Linux, the new path is `~/.cache/JetBrains/<product><version>`.

I will update our bug report guides.

**Comment by Janrupf** [comment]
I did remove the `.git` folder from the hidden files. Also, I can't mark the folder as excluded, so it should probably be ignored entirely, like the commits you added do.

**Comment by m273d15** [comment]
Sorry, the following is a bit out of topic:
> Saros Version: commit 189a144f6f679c891350d2ec94313268aafa4c51 (built from source to enable CLion compat)

@Janrupf I think you already asked for builds that are compatible with CLion (in a former issue). Our nightly builds are compatible with all IntelliJ IDEAs (no restriction enabled). However, it still a nightly build (not stable), but it might be better than building it by yourself.

You can find all builds on the master branch [here](https://github.com/saros-project/saros/actions?query=workflow%3ABuild+branch%3Amaster), you can pick one build like [this](https://github.com/saros-project/saros/actions/runs/80720760).
However, if you pick a build and the "Artifacts" list is empty, you selected a documentation build (due to a technical detail our doc and product build have the same name). Therefore, just pick another build that provides the "Artifacts". Furthermore, make sure that your participants use the same nightly build.

**Comment by tobous** [comment]
Resolved by #926.
